HSEARCH-2192 Highlighting in the Search API / HSEARCH-4815 Better message for FVH being not supported on fields without term vectors / HSEARCH-4822 Extend @FullTextField with `highlightable` configuration (#3418)
hibernate/hibernate-search: documentation/src/main/asciidoc/reference/search-dsl-highlighting.asciidoc

<5> Fetch the results, which will have highlighted fragments.
====

[NOTE]
====
Note that even though the "title" filed in the <<search-dsl-highlighting-using-highlight-projection,example above>>
is a single value field, the resulting projection is a list of strings.
This applies to any highlighting projection.
====

Highlight projections, same as <<search-dsl-projection-field,field projections>> can also be used in a combination with
other projection types as well as with other highlight projections:

[[search-dsl-highlighting-using-highlight-projection-composite]]
.Using a composite highlight projection
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=composite]
----
<1> Start building the query as usual.
<2> Add a regular field projection on a "title" field.
<3> Add a highlight projection on a "description" field.
<4> Note that to get a nonempty list for a highlight projection,
the field we apply such projection to should be a part of a predicate.
If there's no match within the field, we want to highlight - an empty list will be returned by default.

> See <<search-dsl-highlighting-highlighter-options-no-match-size,no match size configuration>> option for more details on how this can be adjusted.

Author: ok,  moved 😃 . I thought that it might be a good place in the composite section since that would probably be a case when more fields/highlights are selected than the predicate is filtering by.

<5> Fetch the results.
====

Sometimes we might want to apply different highlighter types to various fields or specify some non-default configurations for a highlighter.

> To accomplish this, a highlight projection can have an optional `highlighter(..)` parameter set to a name of a named highlighter.

Author: I've added it after highlighter types and before options. Since there's a mention of highlighter types (when we can mix them and when we can't)

As for the highlighter configurations - those can be passed to a query after the where clause.

> There are named highlighters and a query highlighter. Named highlighters must be of the same type as a query one if such was configured.

Author: yeah ... I wanted to have some synonyms so that default is not overused... but on the other hand 😃 as you've been saying it's a ref doc, so better be strict and use the "official names" 😄

For more on tags, see <<search-dsl-highlighting-highlighter-options-tags,the corresponding section>>.

[NOTE]
====
If a full-text field that requires the `Highlightable.FAST_VECTOR` highlighter does not specify the term vectors,
or set them to anything else besides the two mentioned options - then the term vectors for that field
will get "upgraded" to one of `TermVector.WITH_POSITIONS_OFFSETS`/`TermVector.WITH_POSITIONS_OFFSETS_PAYLOADS`
depending on the initially specified value.
====

Picking the highlighter type is the first step in highlighter definition:

.Specifying highlighter type
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=various-highlighter-types]
----
<1> Starting a definition of a default highlighter.
Configuration options are limited to the most common ones that can be found in other highlighters.
<2> Starting a definition of a plain highlighter.
<3> Starting a definition of a unified highlighter.
<4> Starting a definition of a fast vector highlighter.
====

[[search-dsl-highlighting-highlighter-options]]

> == Highlighter options

Author: Ok. I wanted to have a common section for the options so that we could point to it like - "there are various config options see the section and pick whatever you'd like" ...
I guess we can keep the anchor and drop the section. it'll then take the user to the first option

[NOTE]
====

> Any optional highlighter settings can be set in a query highlighter and later overridden in a named highlighter.

Author: This entire note is removed and similar info is added to the named highlighters section.

<2> Passing a collection of open tags and a single closing tag that will be used to highlight the text.
It can be helpful when configuring a tag schema that differs only in an attribute of an opening tag.
<3> Passing a pair of collections containing open/close tags that will be used to highlight the text.
====

Additionally, fast vector highlighter has the option to enable a tag schema and set it to `HighlighterTagSchema.STYLED`
to use a predefined set of tags.

.Setting styled tags schema
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=fvh-schema]
----
<1> Passing a styled tags schema that will be used to highlight the text.
====

Using a styled tags schema is just a shortcut to defining the tags as:

.Setting tags as if the styled tags schema is used
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=fvh-schema-alternative]
----
<1> Passing the same collection of tags used when a styled schema is applied.
====

[NOTE]
====

> If a styled tags schema is applied to a highlighter, then any other tag configurations are ignored, and the schema tags are used.

Author: mmm, so this is a dedicated option for Elasticsearch (`tags_schema`). so we'd need to track what was called last ... 
unless we ignore this dedicated option and just pass the tags schema ourselves ... 
I'll keep it as is for now.

> If a styled tags schema is applied to a highlighter, then any other tag configurations are ignored, and the schema tags are used.

Reviewer: > mmm, so this is a dedicated option for Elasticsearch (tags_schema). so we'd need to track what was called last ...

You can just have the `.tags()` method reset the tag schema to `null`, and vice-versa.

> If a styled tags schema is applied to a highlighter, then any other tag configurations are ignored, and the schema tags are used.

Author: I've removed this note and moved the one about multiple calls to the end of this section.
`SearchHighlighterBuilder` had  a `clearTags()` method that now also clears the tag schema, and Javadocs for the tag methods are updated 😃

No match size configuration allows you to still get some text returned even if the field didn't contribute to the document match,
and there's nothing to be highlighted in it.
The number set by this property defines the number of characters to be included starting at the beginning of a field.
Depending on the highlighter type, the amount of text returned might not precisely match the configured setting since
highlighters usually try not to break the text in the middle of a word/sentence, depending on their configuration.
By default, this setting is set to `0` and text will only be returned if there's something to highlight.

.Setting the no match size
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=no-match-size]
----
<1> We are looking for matches in the title.
<2> In case a word can also be found in the description - we'd want it to be highlighted.
<3> Setting no match size to `100` to still get at least `100` first characters of a description even if the match for
a word we are searching for is not found.
====

[NOTE]
====

> This setting is supported by all highlighter types.

Author: Well... it is supported; it's just that the support is limited 😃. I'll replace this note, since we are dropping those anyway, with the one explaining the limitations of this option with Lucene's unified highlighter.

====

[[search-dsl-highlighting-highlighter-options-fragment-size]]
=== Fragment size and number of fragments
Fragment size limits the amount of text included in each highlighted fragment. Depending on the highlighter type/configuration,
the amount of text per fragment might be more than specified. By default, the fragment size is equal to `100` characters.

A number of fragments configuration sets the maximum number of strings included in the resulting highlighted list.
By default, the number of fragments is equal to `5`.

A combination of these settings can be helpful when highlighting large text fields.

.Setting the fragment size and number of fragments
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=fragment-size]
----
<1> Configuring the fragment size.
<2> Configuring the maximum number of fragments to be returned.
====

[NOTE]
====

> These settings are supported by all highlighter types except for Lucene's default/unified highlighters.

Author: Both plain and fast vector highlighters should be ok in Lucene.

====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=basic-order]
----
<1> A query that would boost the match of one of the searched words.
<2> Configuring the order by the score to be enabled.
====

[NOTE]
====
This setting is supported by all highlighter types.
====

[[search-dsl-highlighting-highlighter-options-fragmenter]]
=== Fragmenter
By default, the plain highlighter breaks up text into same-sized fragments but tries to avoid breaking up a phrase to be highlighted.
This is the behaviour of the `HighlighterFragmenter.SPAN` fragmenter. Alternatively, fragmenter can be set to `HighlighterFragmenter.SIMPLE`
that simply breaks up text into same-sized fragments.

.Setting the fragmenter
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=basic-fragmenter]
----
<1> Configuring the simple fragmenter.
====

[NOTE]
====
This setting is supported only by the plain highlighter.
====

[[search-dsl-highlighting-highlighter-options-boundary-scanner]]
=== Boundary scanner

> Unified and fast vector highlighters use boundary scanners to create highlighted fragments. They use a break iterator
> when looking for a fragment's word/sentence boundaries. An optional locale parameter can be supplied to specify how to search

Reviewer: I doubt that "break iterator" will mean anything to anyone reading this.

```suggestion
Unified and fast vector highlighters use boundary scanners to create highlighted fragments: they try to expand highlighted fragments by scanning text before and after those fragments for word/sentence boundaries.

An optional locale parameter can be supplied to specify how to search
```

for sentence and word boundaries. A sentence boundary scanner is a default option for the unified highlighters.

There are two ways to supply boundary scanner configuration to a highlighter.

.Setting the boundary scanner with DSL
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-dsl]
----
<1> Start the definition of a boundary scanner.

> <2> Pick a boundary scanner type - a word scanner in this case.

Author: Yes!!! 😄😆

<3> Set an optional locale.
<4> End the definition of a boundary scanner.
<5> Set any other configurations.
====

.Setting the boundary scanner using lambda
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-lambda]
----
<1> Pass a lambda configurer to set up a boundary scanner.
<2> Set any other configurations
====

> Additionally, fast vector highlighter can use a character boundary scanner which relies on two other configurations -

Reviewer: ```suggestion
Alternatively, fast vector highlighter can use a character boundary scanner which relies on two other configurations -
```

boundary characters and boundary max scan. When a character boundary scanner is used after a highlighted fragment is formed
with highlighted text centred, the highlighter checks for the first occurrence of any configured boundary characters to the left and right
of a currently created fragment. This lookup happens only for a maximum number of characters configured by the boundary max scan option.

> If boundary characters were not found - no additional text will be included besides the already highlighted phrase

Reviewer: ```suggestion
If no boundary characters are found, no additional text will be included besides the already highlighted phrase
```

with surrounding text based on a fragment size setting. The default list of boundary characters includes `.,!? \t\n`.
The default boundary max scan is equal to `20` characters.

Character boundary scanner is a default option for the fast vector highlighters.

.Setting the character boundary scanner
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-char]
----
Assuming that we have a text that contains multiple paragraphs separated with a new line (`\n`), we want to
get the entire paragraph containing the highlighted phrase. To do so, boundary characters will be set to `\n`, and
the max scan option will be based on the number of characters in paragraphs.

<1> Start the definition of a boundary scanner.

> <2> Pick a boundary scanner type - a character scanner.

Reviewer: ```suggestion
<2> Pick a boundary scanner type -- a character scanner.
```

<3> Set a string containing boundary characters. The overloaded methods can accept a `String` or a `Character` array.
<4> Set the max scan option.
<5> End the definition of a boundary scanner.
<6> Set any other configurations.
====

[NOTE]
====
This setting is supported by the unified and fast vector highlighter types.
====

[[search-dsl-highlighting-highlighter-options-phrase-limit]]
=== Phrase limit
Phrase limit allows to specify the maximum number of matching phrases in a document for highlighting. By default, it is equal to `256`.
This setting can be helpful if a field contains many matches and has a lot of text overall, but we are not interested in

> highlighting every occurrence.

Reviewer: I don't get what's the difference with `.numberOfFragments`?

<4> Add a <<search-dsl-projection-highlight,highlight projection>> on a "description" field.
<5> Provide a predicate to filter documents and to use to highlight the results.
<6> Fetch the results.
====

A highlighter behavior can be configured. See various available <<search-dsl-highlighting-highlighter-options,configuration options>>.
A highlighter definition is provided after a where clause of a query:

[[search-dsl-highlighting-highlighter-confguring-default-highlighter-simple]]
.Configuring a default highlighter
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=very-basic-config]
----
<1> Specify a predicate that would look for matches in both `title` and `description` fields.
<2> Specify the details of the default highlighter. Setting the no match size to a positive value to let the highlighter know
that we want to get some text back even if there will be nothing to highlight in a particular field.
<3> Fetch the results.
====

[[search-dsl-highlighting-highlighter-type]]
== Highlighter type

> Before a highlighter can be configured, you need to pick its type:
> 
> Picking the highlighter type is the first step in a highlighter definition:

Reviewer: ```suggestion
Before a highlighter can be configured, you need to pick its type.
Picking the highlighter type is the first step in a highlighter definition:
```

<2> Starting the definition of a unified highlighter.
<3> Starting the definition of a fast vector highlighter.
====

There are three options to choose from when it comes to the highlighter type:

[[search-dsl-highlighting-highlighter-type-plain]]
Plain::
The plain highlighter can be useful for simple queries targeting a single field on a small number of documents.
This highlighter uses a standard Lucene highlighter. It reads the string value of a highlighted field,
then creates a small in-memory index from it and applies query logic to perform the highlighting.
[[search-dsl-highlighting-highlighter-type-unified]]
Unified::
The unified highlighter is used by default and does not necessarily rely on re-analyzing the text, as it can
get the offsets either from postings or from term vectors.
+
This highlighter uses a break iterator (breaks the text into sentences by default) to break the text into later scored passages.
It better supports more complex queries. Since it can work with prebuilt data, it performs better in case of a larger amount
of documents compared to the plain highlighter.
[[search-dsl-highlighting-highlighter-type-fast-vector]]
Fast vector::

> In addition to using a break iterator similar to the unified highlighter, it can use the boundary characters

Reviewer: It's a bit confusing to use "it" to refer to something that you didn't really name in a previous sentence.

```suggestion
The fast vector highlighter, in addition to using a break iterator similar to the unified highlighter, can use the boundary characters
```

to control the highlighted snippet.
+
This is the only highlighter that can assign different weights to highlighted fragments, allowing it to show
a fragment score differences by wrapping it with a different tag.
For more on tags, see <<search-dsl-highlighting-highlighter-options-tags,the corresponding section>>.

> 
> [NOTE]
> ====
> If a full-text field that requires the `Highlightable.FAST_VECTOR` highlighter does not specify the term vector storage strategy,
> or set it to anything other than `TermVector.WITH_POSITIONS_OFFSETS`/`TermVector.WITH_POSITIONS_OFFSETS_PAYLOADS`,
> then the resulting term vector storing strategy for that field will get "upgraded" to one of `TermVector.WITH_POSITIONS_OFFSETS`/`TermVector.WITH_POSITIONS_OFFSETS_PAYLOADS`
> depending on the initially specified value.
> ====

Reviewer: Maybe this isn't the right place to talk about this? It seems more related to mapping, and I think you already explain this in the mapping section.

```suggestion
```

Alternatively, just mention in the description of the fast vector highlighter that enabling it requires mapping changes, and link to the appropriate section.

[[search-dsl-highlighting-highlighter-named-highlighters]]
== Named highlighters

Sometimes we might want to apply different highlighters to various fields.
We have already seen that a <<search-dsl-highlighting-highlighter-confguring-default-highlighter-simple, highlighter can be configured>>.
The highlighter from that example is called the default highlighter. Search queries also allow to configure named highlighters.
A named highlighter has the same configuration capabilities as the default one.

> Named highlighters override the options set by the default highlighter if such was configured.

Reviewer: It feels weird to switch from singular to plural like this. Maybe:

```suggestion
It overrides the options set by the default highlighter if such was configured.
```

> If a default highlighter was configured for a query then all of named highlighters configured on the same query

Reviewer: ```suggestion
If a default highlighter was configured for a query then every named highlighter configured on the same query
```

the <<search-dsl-projection-highlight,highlight projection definition>>, then that particular highlighter will be applied
to a field projection. Named highlighters can be reused withing a query,
i.e. the same name of a named highlighter can be passed to multiple highlight projections.

.Configuring both default and named highlighters
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=basic-config]
----
<1> Add the highlight projection on a "title" field. This projection uses the default highlighter.
<2> Add the highlight projection on a "description" field and specify the name of the named highlighter.
<3> Specify the details of the default highlighter.
<4> Specify the details of the named highlighter. Note that the name matches the name passed to the configuration of the "description"
highlight projection.
<5> Fetch the results.
====

[NOTE]
====
The name of a named highlighter cannot be `null` or an empty string. An exception will be thrown if such values are used.
====

> [[search-dsl-highlighting-highlighter-options]][[search-dsl-highlighting-highlighter-options-tags]]
> == Tags

Reviewer: This doesn't render correctly.

```suggestion
[[search-dsl-highlighting-highlighter-options-tags]]
== [[search-dsl-highlighting-highlighter-options]] Tags
```

to use a predefined set of tags.

.Setting a styled tags schema
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=fvh-schema]
----
<1> Passing a styled tags schema that will be used to highlight the text.
====

Using a styled tags schema is just a shortcut to defining tags as:

.Setting tags as if the styled tags schema is used
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=fvh-schema-alternative]
----
<1> Passing the same collection of tags used when a styled schema is applied.
====

[NOTE]
====
If a styled tags schema is applied to a highlighter, then any other tag configurations are ignored, and the schema tags are used.
====

[[search-dsl-highlighting-highlighter-options-encoder]]
== Encoder
Encoding can be applied to the highlighted snippets when highlighting the fields that store HTML. Applying an
HTML encoder to a highlighter will encode the text for inclusion into an HTML document:

> it will replace HTML meta-characters such as `<` with their entity equivalent such as `&amplt;`;

Reviewer: For some reason this triggers a build error when generating PDF documentation (`-Pdocumentation-pdf`).

I think there's a missing `;`?

```suggestion
it will replace HTML meta-characters such as `<` with their entity equivalent such as `&amp;lt;`;
```

Alternatively, maybe you should use the `+` notation:

```suggestion
it will replace HTML meta-characters such as `<` with their entity equivalent such as `+&lt;+`;
```

Depending on the highlighter type, the amount of text returned might not precisely match the configured value since
highlighters usually try not to break the text in the middle of a word/sentence, depending on their configuration.
By default, this option is set to `0` and text will only be returned if there's something to highlight.

.Setting the no match size
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=no-match-size]
----
<1> We are looking for matches in the title.
<2> In case a word can also be found in the description, we'd want it to be highlighted.
<3> Setting the no match size to `100` to still get at least `100` first characters of a description even if the match for
a word we are searching for is not found.
====

[NOTE]
====
The unified highlighter from the <<backend-lucene,Lucene backend>> has a limited support for this option.
It cannot limit the amount of the returned text, and works more like a boolean flag to enable/disable the
feature. If a highlighter of this type has the option not set or set to `0` then no text is returned when

> there was no match found. Otherwise, if the option for a highlighter of this type was set to a positive integer - all

Reviewer: ```suggestion
there was no match found. Otherwise, if the option for a highlighter of this type was set to a positive integer, all
```

.Setting the character boundary scanner
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-char]
----
Assuming that we have a text that contains multiple paragraphs separated with a new line (`\n`), we want to
get the entire paragraph containing the highlighted phrase. To do so, boundary characters will be set to `\n`, and
the max scan option will be based on the number of characters in paragraphs.

<1> Start the definition of a boundary scanner.
<2> Pick a boundary scanner type -- a character scanner.
<3> Set a string containing boundary characters. The overloaded methods can accept a `String` or a `Character` array.
<4> Set the max scan option.
<5> End the definition of a boundary scanner.
<6> Set any other options.
====

[NOTE]
====
This option is supported by the unified and fast vector highlighter types.
====

[[search-dsl-highlighting-highlighter-options-phrase-limit]]
== Phrase limit
Phrase limit allows to specify the maximum number of matching phrases in a document for highlighting.

> The highlighter will be going through the text and as soon as it performs the phrase limit number of highlights it'll stop
> leaving any further occurrences as not highlighted.

Reviewer: Can you document the difference between this and the "number of fragments" above? You explained it in a comment, but not in the documentation:

> mmm, so the number of fragments is the number of items in the returned list. but this phrase limit - is how many phrases will get "tagged" so if we have a text with a word that is often repeated and we search for it - highlighter will wrap that word for `Phrase limit` times and all of them might be in the first fragment or it may span between all the fragments.
> 
> here's an example where we searched for a `bank`:
> 
> ```
> // note that there's a second "bank" occurrence, but since we only wanted one phrase highlighted it is ignored:
> contacts a police inspector and the manager of a nearby <em>bank</em>. With Watson, they hide in the bank vault and catch
> ```
> 
> the second bank is ignored because the phrase limit was set to 1.

Maybe something like this:

```suggestion
The highlighter will be going through the text and as soon as it reaches the maximum number of highlighted phrases it'll stop leaving any further occurrences as not highlighted.

[NOTE]
====
This limit is different from the <<search-dsl-highlighting-highlighter-options-fragment-size,maximum number of fragments>>:

1. Fragments are the strings returned by the highlight projections, while phrases are the sequences of highlighted (matching) terms in each fragment. A fragment may include multiple highlighted phrases, and a given phrase may appear in multiple fragments.
2. The phrase limit is about limiting highlighting of multiple occurrences of a *same phrase*, so with a phrase limit of `1` the phrase `shiny robot` will only be highlighted once, but there may be other highlighted phrases such as `great detective` (also highlighted only once).
==== 
```

Is this correct?

Author: Almost, I think there's a single "counter", so only `shiny robot` will get highlighted... I'll update the `2` point to:

2. The phrase limit is about limiting the highlighting of occurrences of matched phrases, be it multiple occurrences of the same phrase or a mix of different phrases.
For example, if we were to search for `fox` and `dog` occurrences in the sentence `The quick brown fox jumps over the lazy dog` and set the phrase limit to `1`, then we'll have only `fox` being highlighted since it was the first match in the text and the phrase limit was reached.
